torch_geometric/data/materialized_graph.py

```diff
@@ -0,0 +1,125 @@
+
+    # Core ####################################################################
+
+    @abstractmethod
+    def _put_edge_index(self, edge_index: EdgeTensorType,
+                        edge_attr: EdgeAttr) -> bool:
+        return None
+
+    def put_edge_index(self, edge_index: EdgeTensorType, *args,
+                       **kwargs) -> bool:
+        r"""Synchronously adds an edge_index tensor to the materialized graph.
+
+        Args:
+            tensor(EdgeTensorType): an edge_index in a format specified in
+            attr.
+            **attr(EdgeAttr): the edge attributes.
+        """
+        edge_attr = self._edge_attr_cls.cast(*args, **kwargs)
+        if not edge_attr.layout:
+            raise AttributeError(
+                "An edge layout is required to store an edge index, but one "
+                "was not provided.")
+
+        # NOTE implementations should take care to ensure that `SparseTensor`
+        # objects are treated properly here.
+        return self._put_edge_index(edge_index,
+                                    self._edge_attr_cls.cast(*args, **kwargs))
+
+    @abstractmethod
+    def _get_edge_index(self, edge_attr: EdgeAttr) -> EdgeTensorType:
+        return None
+
+    def get_edge_index(self, *args, **kwargs) -> EdgeTensorType:
+        r"""Synchronously gets an edge_index tensor from the materialized
+        graph.
+
+        Args:
+            **attr(EdgeAttr): the edge attributes.
+
+        Returns:
+            EdgeTensorType: an edge_index tensor corresonding to the provided
+            attributes, or None if there is no such tensor.
+        """
+        return self._get_edge_index(self._edge_attr_cls.cast(*args, **kwargs))
+
+    def sample(
+        self,
+        input_nodes: Union[LongTensor, Dict[EdgeType, LongTensor]],
+        *args,
+        **kwargs,
+    ) -> None:
+        r"""Samples the materialized graph to obtain a sampled subgraph,
+        utilizing sampling routines provided by `torch_sparse`.
+
+        Args:
+            input_nodes: the input nodes to sample from, either provided as a
+                :class:`torch.LongTensor` or a dictionary mapping node type to
+                a list of input nodes.
+
+        Notes:
+            Most sampling methods operate efficiently on sorted edge indices.
+            This class makes no guarantees on the formats of the edge indices
+            that are passed as input. Implementations should account for this
+            (e.g. by converting all edge indices to CSC) before sampling.
+        """
+        # NOTE for the future: this is expected to call a C++ implementation
+        # in pyg-lib.
+        return None
+
+    # Python built-ins ########################################################
+
+    def __setitem__(self, key: EdgeAttr, value: EdgeTensorType):
+        key = self._edge_attr_cls.cast(key)
+        self.put_edge_index(value, key)
+
+    def __getitem__(self, key: EdgeAttr) -> EdgeTensorType:
```

Comment:
How do these methods relate to `BaseData` and `FeatureStore`?

Reply:
They are overridden by `BaseData` and its implementations. I think that both `FeatureStore` and `MaterializedGraph` should define these dunder methods internally as both a feature store and materialized graph can be used independently of one another, but implementations that combine the two will need to intelligently override them or deal with some nasty bugs.